### DISCUSSION TITLE
Look into zero heap for embedded applications

### DISCUSSION THREAD

**Discussion by stephenberry** [stephenberry/glaze#121] [Ideas]
Likely have to rework error handling to avoid exceptions in this context.

**Comment by mwalcott3** [comment]
Also, static thread_local std::string temp buffers will need to get replaced. We could do this fairly easily if a max key size was supplied as a compile-time option. Perhaps there is an even better solution. I'm not very familiar with embedded programming patterns/solutions.

**Comment by stephenberry** [comment]
I did update the code so that the `static thread_local std::string` is only needed for non-contiguous buffers or keys with escapes. So, in an embedded context a contiguous buffer can be used and we just wouldn't allow keys with escaped characters.

**Comment by stephenberry** [comment]
We are continuing to move towards using the heap only when necessary. We are using glaze on embedded platforms and will continue to improve its embedded support. Closing this now since we are intending to only allocate where needed and continuing to make progress.

**Comment by sanfusu** [comment]
Some embedded platforms do not support full thread functionality (they do not save the thread pointer register during context switches). Therefore, adding additional compilation options to disable the use of thread_local may deliver better compatibility.
